import "ent/acoes.proto";
import "ent/aux.proto";

package ent;
option java_package = "com.matferib.Tabuleiro.ent";

enum TipoEntidade {
  TE_ENTIDADE = 0;
  TE_FORMA = 1;
  TE_COMPOSTA = 10;
}

// A ordem dos valores da enum deste proto eh importante, porque eh usada para construir o menu de desenho.
enum TipoForma {
  TF_CILINDRO = 0;
  TF_CIRCULO = 1;
  TF_CONE = 2;
  TF_CUBO = 3;
  TF_ESFERA = 4;
  TF_LIVRE = 5;
  TF_PIRAMIDE = 6;
  TF_RETANGULO = 7;
  TF_TRIANGULO = 8;
}

enum TamanhoEntidade {
  TM_MINUSCULO = 0;
  TM_DIMINUTO = 1;
  TM_MIUDO = 2;
  TM_PEQUENO = 3;
  TM_MEDIO = 4;
  TM_GRANDE = 5;
  TM_ENORME = 6;
  TM_IMENSO = 7;
  TM_COLOSSAL = 8;
}

enum ResultadoSalvacao {
  RS_FALHOU = 0;
  RS_MEIO = 1;
  RS_QUARTO = 2;
  RS_ANULOU = 3;
}

// Informacoes sobre a classe de um personagem.
message InfoClasse {
  optional string id = 1;
  optional int32 nivel = 2;
  optional int32 nivel_conjurador = 3;
  // Bonus base de ataque do nivel.
  optional int32 bba = 4;
  // Exemplo: modificador de sabedoria para clerigos.
  optional int32 modificador_atributo_conjuracao = 5;
}

message EntidadeProto {
  // Tipo da entidade, determinada pela classe folha. Nao deve ser alterado.
  optional TipoEntidade tipo = 1;
  optional TipoForma sub_tipo = 18;
  // Identificador global da entidade composto pelo id_criador << 28 | id_local.
  // Portanto, o maximo de jogadores é 16. É importante que o identificador da entidade caiba em
  // um GLuint para permitir realização da operação de picking no 3d.
  optional uint32 id = 2;
  // Pontos de vida da entidade.
  optional int32 pontos_vida = 11;
  // Pontos de vida temporarios.
  optional int32 pontos_vida_temporarios = 59;
  // Maximo de pontos de vida da entidade.
  optional int32 max_pontos_vida = 12;
  // Exemplo: 4d8 + 8.
  optional string dados_vida = 19;
  // Para os jogadores poderem se identificar.
  optional string rotulo = 20;
  // Informacoes adicionais sobre a entidade. Visivel apenas para o mestre ou se a entidade puder ser selecionada.
  // Algumas informacoes serao concatenadas ao valor do campo.
  repeated string rotulo_especial = 30;
  // Os varios tamanhos de uma entidade no tabuleiro.
  optional TamanhoEntidade tamanho = 3 [default=TM_MEDIO];
  // Se verdadeiro, objeto sera desenhado de forma achatada.
  optional bool achatado = 17 [default=false];
  // Posicao atual da entidade. Note que o Z eh quanto a criatura esta acima da translacao. A real posicao dela
  // esta em pos.z + translacao.z.
  optional Posicao pos = 4;
  // Para formas (exceto translacao).
  repeated Posicao ponto = 21;
  optional Escala escala = 22;
  optional float rotacao_z_graus = 23;  // rotacao do objeto ao redor de Z em graus.
  optional float rotacao_y_graus = 29;  // rotacao do objeto ao redor de Y em graus.
  optional float rotacao_x_graus = 33;  // rotacao do objeto ao redor de X em graus.
  optional float translacao_z_deprecated = 24 [deprecated=true];
  repeated EntidadeProto sub_forma = 25;  // Para entidades compostas.
  enum TipoTransicao {
    TRANS_NENHUMA = 0;
    // Acao na entidade transita cenario.
    TRANS_CENARIO = 1;
    // Acao na entidade recolhe tesouro.
    TRANS_TESOURO = 2;
  }
  optional TipoTransicao tipo_transicao = 61;
  // Algumas formas permitem a transicao de um cenario para outro. Quando presente, esta forma tera a capacidade de transitar
  // entidades e a camera para outro cenario. Valor de -1 eh usado para cenario principal e -2 eh reservado para limpar a transicao.
  optional Posicao transicao_cenario = 35;

  // Destino da entidade.
  optional Posicao destino = 5;
  // Cor da entidade.
  optional Cor cor = 6;
  // Iluminação da entidade.
  optional IluminacaoPontual luz = 7;
  // Textura.
  optional InfoTextura info_textura = 8;
  // Modelo 3d.
  optional Modelo3d modelo_3d = 34;

  // BITS DAS ENTIDADE.
  optional int32 aura = 9 [deprecated=true];
  optional float aura_m = 41;
  // Visibilidade. Apenas o dono pode ver.
  optional bool visivel = 10 [default=true];
  // Verdadeiro se a entidade for voadora.
  optional bool voadora = 13 [default=false];
  optional float z_antes_voo = 49;
  // Entidade esta caida.
  optional bool caida = 14 [default=false];
  // Entidade esta morta.
  optional bool morta = 15 [default=false];
  // Se false, apenas o mestre pode ver propriedades da entidade, seleciona-la e fazer picking.
  optional bool selecionavel_para_jogador = 16 [default=true];
  // Se verdadeiro, o mestre nao conseguira selecionar a entidade. Ela sera apenas editavel com duplo clique.
  optional bool fixa = 32 [default=false];
  // Se falso, a entidade nao faz sombra.
  optional bool faz_sombra = 39 [default=true];
  // Tipo de visao da entidade.
  optional TipoVisao tipo_visao = 36;
  // Alcance da visao no escuro em metros.
  optional float alcance_visao = 37;
  // Entidade causa colisao.
  optional bool causa_colisao = 48;
  // Indica que a entidade esta apoiada.
  optional bool apoiada = 50 [default=true];
  // Indica que a entidade esta espiando de um lado ou do outro. -1 esquerda, 1 direita.
  optional int32 espiando = 58;
  // Se verdadeiro, desligara luz para desenho.
  optional bool ignora_luz = 75;
  // Se verdadeiro, desligara backface culling.
  optional bool dois_lados = 76;

  // TODO implementar as coisas de surpresa.
  optional bool surpreso = 64;
  optional bool furtivo = 62;

  // Ultima acao para display. O valor eh um id no mapa de acoes do tabuleiro.
  optional string ultima_acao = 26;
  // Ultimas acoes executadas da unidade. Funciona como uma fila onde o indice 0 eh a acao mais recente e o
  // tamanho maximo eh dado pela constante MaxNumAcoes.
  repeated string lista_acoes = 38;
  // Resultado da proxima salvacao da entidade.
  optional ResultadoSalvacao proxima_salvacao = 27;
  // A direcao do ultimo impacto sofrido pela entidade.
  optional Posicao direcao_queda = 28;
  // Evento de entidade. A cada rodada, o contador de rodadas diminuira ate zero.
  // Ao termino do contador, o tabuleiro sinalizara a entidade.
  message Evento {
    optional int32 rodadas = 1;
    optional string descricao = 2;
    optional int32 id_efeito = 4;    // Se o evento for reconhecido como algum efeito, o id estara aqui. Ids validos em nt/util.h.
    optional int32 complemento = 3;  // Alguns efeitos tem complementos.
  }
  repeated Evento evento = 31;

  // Dados para ataque automatico.
  message DadosAtaque {
    // Deve casar com as strings do arquivo de dados das acoes.
    optional string rotulo = 9;
    optional string tipo_ataque = 1;
    optional int32 bonus_ataque = 2;
    // TODO string de critico. Parseavel.
    optional string dano = 3;
    optional int32 margem_critico = 6 [default=20];
    optional int32 multiplicador_critico = 7 [default=2];
    optional int32 ca_normal = 4;
    optional int32 ca_surpreso = 5;
    optional int32 ca_toque = 8;
    optional float alcance_m = 10;
    optional int32 incrementos = 11;  // Quantos incrementos de distancia sao permitidos.
    optional AcaoProto acao = 12;     // Para fazer merge com a acao do tipo_ataque.
  }
  repeated DadosAtaque dados_ataque = 44;
  // Independe de ataque especifico.
  message DadosAtaqueGlobal {
    optional string dano_furtivo = 1;
    // Ver funcao ModificadorAtaque em util.
    // Alguns bonus de ataque genericos.
    optional bool ataque_menos_1 = 2;
    optional bool ataque_menos_2 = 3;
    optional bool ataque_menos_4 = 4;
    optional bool ataque_menos_8 = 14;
    optional bool ataque_mais_1 = 5;
    optional bool ataque_mais_2 = 6;
    optional bool ataque_mais_4 = 7;
    optional bool ataque_mais_8 = 15;
    // Dano generico.
    optional bool dano_menos_1 = 8;
    optional bool dano_menos_2 = 9;
    optional bool dano_menos_4 = 10;
    optional bool dano_menos_8 = 19;
    optional bool dano_mais_1 = 11;
    optional bool dano_mais_2 = 12;
    optional bool dano_mais_4 = 13;
    optional bool dano_mais_8 = 16;
    optional bool dano_mais_16 = 17;
    optional bool dano_mais_32 = 18;
  }
  optional DadosAtaqueGlobal dados_ataque_globais = 73;

  // Dados de defesa global (independe de ataque).
  message DadosDefesa {
    optional string tipo_defesa = 1 [deprecated=true];
    optional int32 classe_armadura = 2 [deprecated=true];
    optional bool imune_critico = 3;
  }
  optional DadosDefesa dados_defesa = 45;

  // Geracao de tesouro automatica?
  message DadosTesouro {
    optional string tesouro = 1;
  }
  optional DadosTesouro tesouro = 46;
  // Dados de iniciativa da entidade.
  optional int32 iniciativa = 55;
  optional int32 modificador_iniciativa = 56;
  // Algumas entidades tem varias formas.
  repeated EntidadeProto outras_formas = 60;

  // Informacoes sobre a classe do personagem.
  repeated InfoClasse info_classes = 63;
  // Mesmo que haja um modelo 3d, desenha o tijolo de base da entidade.
  optional bool desenha_base = 65;
}
